hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/ServerCrashProcedure.java

```diff
@@ -466,20 +466,31 @@ private void assignRegions(MasterProcedureEnv env, List<RegionInfo> regions) thr
       RegionStateNode regionNode = am.getRegionStates().getOrCreateRegionStateNode(region);
       regionNode.lock();
       try {
+        // This is possible, as when a server is dead, TRSP will fail to schedule a RemoteProcedure
+        // to us and then try to assign the region to a new RS. And before it has updated the region
+        // location to the new RS, we may have already called the am.getRegionsOnServer so we will
+        // consider the region is still on us. And then before we arrive here, the TRSP could have
+        // updated the region location, or even finished itself, so the region is no longer on us
+        // any more, we should not try to assign it again. Please see HBASE-23594 for more details.
+        if (!serverName.equals(regionNode.getRegionLocation())) {
```

Comment:
In

HBASE-23369 Auto-close 'unknown' Regions reported as OPEN on RegionServers

... there is similar change but in HBCKSCP where we check RegionStateNode location against current ServerName before acting... skipping if they don't match. It is more paranoid (overly so now on second reading). It does...

` +      } else if (!rs.isOnServer(getServerName())) {
 +        if (((RegionState) rs).isOpened()) {
 +          // OPEN on another server altogether! If we queue here, it'll be double assign.
 +          // Why is this? Should SCP update Region state incrementally instead of one lump on end?
 +          LOG.warn("{} is OPEN and NOT on {}: Skipping re-assign.", rs, getServerName());
 +          continue;
 +        }`

You should probably do null check here just in case.

Reply:
And this fix is better than what HBASE-23369 tries to do since it catches the problem before damage is done.

Reply:
What null check? Here serverName and regionNode will never be null? And regionNode.getRegionLocation could return null and that's why I use serverName.equals instead of reionNode.getRegionLocation.equals...

Reply:
Was thinking check for null location but as you note, ServerName#equals takes care of it. Let me resolve this comment.